src/main.cpp

```diff
@@ -55,40 +55,52 @@ class any_error_reporter {
     std::visit(
         [&](auto &r) {
           using reporter_type = std::decay_t<decltype(r)>;
-          if constexpr (std::is_base_of_v<vim_qflist_json_error_reporter,
+          if constexpr (std::is_base_of_v<error_tape<vim_qflist_json_error_reporter>,
                                           reporter_type>) {
-            r.set_source(input, file.path, file.vim_bufnr);
+            r.get_reporter()->set_source(input, file.path, file.vim_bufnr);
           } else {
-            r.set_source(input, file.path);
+            r.get_reporter()->set_source(input, file.path);
           }
         },
-        this->reporter_);
+        this->tape_);
   }
 
   error_reporter *get() noexcept {
-    return std::visit([](error_reporter &r) { return &r; }, this->reporter_);
+    try {
+      return &std::get<error_tape<text_error_reporter>>(this->tape_);
+    } catch(std::bad_variant_access&) {
+      return &std::get<error_tape<vim_qflist_json_error_reporter>>(this->tape_);
+    }
+  }
+
+  bool get_error() noexcept {
+    try {
+      return std::get<error_tape<text_error_reporter>>(this->tape_).get_error();
+    } catch(std::bad_variant_access&) {
+      return std::get<error_tape<vim_qflist_json_error_reporter>>(this->tape_).get_error();
+    }
```

Comment:
Same with `get`; use `std::visit` if possible.

Reply:
I would like to use `std::visit`.
But the code would be:
`return std::visit([](error_tape<> &r) { return r.get_error(); }, this->tape_);`
And this does not compile because error_tape needs a type as template and I do not see which type to use...

Reply:
Ah, right. And `return std::visit([](error_reporter &r) { return r.get_error(); }, this->tape_);` won't work because `error_reporter::get_error` isn't a thing.

Making the lambda a template should work though:
```cpp
return std::visit([](auto &r) { return r.get_error(); }, this->tape_);
```

Reply:
This is now corrected!

I should have think to use `auto`... This is when you write code like that you understand the ++ of C++ compared to C :laughing:.